tensorflow/compiler/tf2tensorrt/utils/trt_shape_optimization_profiles.cc

```diff
@@ -25,26 +25,109 @@ limitations under the License.
+  *min = dimvec;
+  for (auto& dim : *min) {
+    dim.d[0] = 1;  // Set min batch size to 1.
+  }
+  *opt = dimvec;
+  *max = dimvec;
+}
+
+void TrtShapeOptimizationProfile::ImplicitBatchModeCompatibleStrategy() {
   for (auto& shape_vec : input_shapes_) {
     if (!shape_vec.empty()) {
-      std::vector<nvinfer1::Dims> dimvec(shape_vec.size());
-      absl::c_transform(shape_vec, dimvec.begin(), [](TensorShape shape) {
-        return TensorShapeToTrtDims(shape, false);
-      });
-      // Set min=opt=max.
-      OptimizationProfileConfig profConfig{dimvec, dimvec, dimvec};
+      std::vector<nvinfer1::Dims> dimvec = GetDimVec(shape_vec);
+      std::vector<nvinfer1::Dims> min, opt, max;
+      SetImplicitBatchModeCompatibleProfile(dimvec, &min, &opt, &max);
+      OptimizationProfileConfig profConfig{min, opt, max};
       profiles_.push_back(std::move(profConfig));
-      VLOG(1) << "Created profile " << profiles_.back().DebugString();
+    }
+  }
+}
+
+void TrtShapeOptimizationProfile::OptimalStrategy() {
+  for (auto& shape_vec : input_shapes_) {
+    if (!shape_vec.empty()) {
+      std::vector<nvinfer1::Dims> min = GetDimVec(shape_vec);
+      std::vector<nvinfer1::Dims> opt = min;
+      std::vector<nvinfer1::Dims> max = min;
+      OptimizationProfileConfig profConfig{min, opt, max};
+      profiles_.push_back(std::move(profConfig));
+    }
+  }
+}
+
+// Adjust shape value profile to prevent TRT from removing shape value input
+// bindings whose value is redundant (only a single value matches the profile).
+// This should be removed once the NVIDIA bug 3153064 is fixed.
+Status FixShapeValueProfile(OptimizationProfileConfig* prof,
+                            const std::vector<bool>& is_shape_tensor) {
+  for (int i = 0; i < prof->min.size(); i++) {
+    if (is_shape_tensor[i] &&
+        std::equal(prof->min[i].d, prof->min[i].d + prof->min[i].nbDims,
+                   prof->max[i].d)) {
+      VLOG(2) << "Adjust profile for shape value tensor " << i;
+      prof->max[i].d[0]++;
+    }
+  }
+}
+
+void TrtShapeOptimizationProfile::InitProfiles(
+    const std::vector<PartialTensorShape>& input_partial_shapes) {
+  if (input_shapes_.size() == 0) {
+    VLOG(1) << "Not creating profiles without input_shapes. "
+               "You have to enable profile generation mode first (build).";
+    return;
+  }
+  switch (strategy_) {
+    case ProfileStrategy::kImplicitBatchModeCompatible:
+      VLOG(1) << "Creating profiles with ImplicitBatchModeCompatible strategy";
+      ImplicitBatchModeCompatibleStrategy();
+      break;
+    case ProfileStrategy::kOptimal:
+      VLOG(1) << "Creating profiles with Optimal strategy";
+      OptimalStrategy();
+      break;
+  }
+  // Define a mask that describe which input could be a shape tensor. Note that
+  // here we can have false positives. The shape tensor mask will be updated
+  // once the network is constructed.
+  SetShapeTensorMask(input_partial_shapes);
```

Comment:
SetShapeTensorMask returns "Status" currently. I think it can return void, please check the definition of the routine.
Otherwise, need to handle the return value here.